### PR TITLE
Fix inconsistent snapshot during concurrent update

### PR DESCRIPTION
**Pull request by xiong-gang**
After enabling the global deadlock detector, we can support concurrent updates.
When updating one tuple at the same time, the conflict and wait are moved from
QD to QE. We need to make sure the implicated transaction order on the segments
is also considered on the master when taking the distributed snapshots.

This is reported: https://github.com/greenplum-db/gpdb/issues/9407

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/storage/lmgr/lmgr.c [hunk 1]
```diff
@@ -703,6 +718,45 @@ XactLockTableWait(TransactionId xid, Relation rel, ItemPointer ctid,
...
+void
+GxactLockTableInsert(DistributedTransactionId gxid)
+{
+	LOCKTAG		tag;
+
+	Assert(Gp_role == GP_ROLE_DISPATCH);
+
+	SET_LOCKTAG_DISTRIB_TRANSACTION(tag, gxid);
+
+	(void) LockAcquire(&tag, ExclusiveLock, false, false);
```
Comment (zhangh43) [inline src/backend/storage/lmgr/lmgr.c]: When will this exclusive lock be freed?  do we need something like GxactLockTableDelete

File: src/backend/cdb/cdbtm.c [hunk 2]
```diff
@@ -1231,8 +1241,23 @@ doDispatchDtxProtocolCommand(DtxProtocolCommand dtxProtocolCommand,
 		}
 	}
 
+	MyTmGxactLocal->waitGxids = NULL;
```
Comment (zhangh43) [inline src/backend/cdb/cdbtm.c]: MyTmGxactLocal->waitGxids has waitGxids allocated in TopTransactionContext, so we'd better free it when it's not NIL.

File: src/backend/storage/ipc/procarray.c [hunk 3]
```diff
@@ -4896,6 +4897,39 @@ GetPidByGxid(DistributedTransactionId gxid)
...
+LocalXidGetDistributedXid(TransactionId xid)
+{
+	int index;
+	DistributedTransactionTimeStamp tstamp;
+	DistributedTransactionId gxid = InvalidDistributedTransactionId;
+	ProcArrayStruct *arrayP = procArray;
+
+	LWLockAcquire(ProcArrayLock, LW_SHARED);
+	for (index = 0; index < arrayP->numProcs; index++)
+	{
```
Comment (zhangh43) [inline src/backend/storage/ipc/procarray.c]: ProcArrayLock in this code block could race fiercely. Is there any way we could store this <xid,gxid> pair in advance?

File: src/backend/cdb/cdbtm.c (hunk 2 above)
Comment (xiong-gang) [reply src/backend/cdb/cdbtm.c]: Fixed

File: src/backend/cdb/cdbtm.c [hunk 4]
```diff
@@ -1231,8 +1241,23 @@ doDispatchDtxProtocolCommand(DtxProtocolCommand dtxProtocolCommand,
...
+		{
+			PQclear(result);
+			continue;
+		}
+
+		oldContext = MemoryContextSwitchTo(TopTransactionContext);
+		for (idx = 0; idx < result->nWaits; idx++)
+			MyTmGxactLocal->waitGxids = list_append_unique_int(MyTmGxactLocal->waitGxids, result->waitGxids[idx]);
+		MemoryContextSwitchTo(oldContext);
+	}
```
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/cdb/cdbtm.c]: Fixed

File: src/backend/cdb/cdbtm.c [hunk 5]
```diff
@@ -1907,7 +1933,22 @@ performDtxProtocolPrepare(const char *gid)
...
+	StringInfoData buf;
+	int len = list_length(waitGxids);
 
+	if (len == 0)
+		return;
+
+	pq_beginmessage(&buf, 'w');
+	pq_sendint(&buf, len, 4);
+	foreach(lc, waitGxids)
+		pq_sendint(&buf, lfirst_oid(lc), 4);
```
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/cdb/cdbtm.c]: Done

File: src/backend/storage/ipc/procarray.c (hunk 3 above)
Comment (xiong-gang) [reply src/backend/storage/ipc/procarray.c]: I remember we used to have such kind of structure and we removed it. I did some benchmark and it shows no performance downgrade. I guess we can fix this problem and benchmark the lock contention to see if we need to optimize it...

File: src/backend/storage/lmgr/lmgr.c (hunk 1 above)
Comment (xiong-gang) [reply src/backend/storage/lmgr/lmgr.c]: I think it goes with LockReleaseAll. XactLockTableDelete is used for sub-transactions and we don't have sub-transaction for the global transactions.

File: src/backend/cdb/cdbtm.c [hunk 6]
```diff
@@ -1126,6 +1137,17 @@ isMppTxOptions_ExplicitBegin(int txnOptions)
...
  */
+static int
+compare_int(const void *va, const void *vb)
+{
+	int			a = *((const int *) va);
+	int			b = *((const int *) vb);
+
+	if (a == b)
+		return 0;
+	return (a > b) ? 1 : -1;
```
Comment (pivotal-ning-yu) [inline src/backend/cdb/cdbtm.c]: We can simply replace the above three lines with `return a - b`, there is no need to unify the return value.

File: src/test/isolation2/sql/concurrent_update.sql [hunk 7]
```diff
@@ -10,5 +10,33 @@ INSERT INTO t_concurrent_update VALUES(1,1,'test');
...
+INSERT INTO t_concurrent_update VALUES(1,1);
+
+1: BEGIN;
+1: UPDATE t_concurrent_update SET b=b+10 WHERE a=1;
+2: BEGIN;
+-- transaction 2 will wait transaction 1 on the segment
+2&: UPDATE t_concurrent_update SET b=b+10 WHERE a=1;
+-- transaction 1 suspend before commit, but it will wake up transaction 2 on segment
+1: select gp_inject_fault('before_xact_end_procarray', 'suspend', '', '', '', 1, 1, 0, 1);
+1&: END;
```
Comment (pivotal-ning-yu) [inline src/test/isolation2/sql/concurrent_update.sql]: We'd better put a call to `gp_wait_until_triggered_fault()` to ensure the suspending, otherwise there might be a chance that con2 `END` before con1 on the segment.

Comment (pivotal-ning-yu) [reply src/test/isolation2/sql/concurrent_update.sql]: BTW, we could use the short form of the fault injector, `gp_inject_fault(faultname text, type text, db_id int4)`, the other fields are filled with the same values with your long form.

And in new code it's recommended to call `gp_inject_fault()` like this:
```sql
1: select gp_inject_fault('before_xact_end_procarray', 'suspend', dbid) from gp_segment_configuration where content=0 and role='p';
```

File: src/backend/cdb/cdbtm.c [hunk 8]
```diff
@@ -1232,7 +1257,46 @@ doDispatchDtxProtocolCommand(DtxProtocolCommand dtxProtocolCommand,
...
+	if (totalWaits > 0)
+		waitGxids = palloc0(sizeof(int) * totalWaits);
+
+	totalWaits = 0;
+	for (i = 0; i < resultCount; i++)
+	{
+		int idx;
+		struct pg_result *result = results[i];
+
+		for (idx = 0; idx < result->nWaits; idx++)
```
Comment (kainwen) [inline src/backend/cdb/cdbtm.c]: memcpy here?

File: src/test/isolation2/sql/concurrent_update.sql (hunk 7 above)
Comment (xiong-gang) [reply src/test/isolation2/sql/concurrent_update.sql]: > We'd better put a call to `gp_wait_until_triggered_fault()` to ensure the suspending, otherwise there might be a chance that con2 `END` before con1 on the segment.

It triggers the injector when executing `END` and then block, I don't think we got the chance to call `gp_wait_until_triggered_fault`

Comment (xiong-gang) [reply src/test/isolation2/sql/concurrent_update.sql]: > BTW, we could use the short form of the fault injector, `gp_inject_fault(faultname text, type text, db_id int4)`, the other fields are filled with the same values with your long form.
> 
> And in new code it's recommended to call `gp_inject_fault()` like this:
> 
> ```sql
> 1: select gp_inject_fault('before_xact_end_procarray', 'suspend', dbid) from gp_segment_configuration where content=0 and role='p';
> ```

Since the other session will also hit the injector `before_xact_end_procarray` and we don't want it get suspended, we need to specify the number of occurrences, so can't use the simplified version

File: src/backend/cdb/cdbtm.c (hunk 6 above)
Comment (xiong-gang) [reply src/backend/cdb/cdbtm.c]: The other qsort comparator also returns the same -1, 1 and 0 value, I think it's better to keep it the same.

File: src/backend/access/transam/distributedlog.c [hunk 9]
```diff
@@ -423,6 +423,29 @@ DistributedLog_SetCommittedTree(TransactionId xid, int nxids, TransactionId *xid
...
+{
+	int			page = TransactionIdToPage(localXid);
+	int			entryno = TransactionIdToEntry(localXid);
+	int			slotno;
+	DistributedLogEntry *ptr;
+
+	Assert(!IS_QUERY_DISPATCHER());
+
+	slotno = SimpleLruReadPage_ReadOnly(DistributedLogCtl, page, localXid);
+	ptr = (DistributedLogEntry *) DistributedLogCtl->shared->page_buffer[slotno];
```
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/access/transam/distributedlog.c]: no, here's only distributed log.

File: src/backend/cdb/cdbtm.c (hunk 8 above)
Comment (xiong-gang) [reply src/backend/cdb/cdbtm.c]: it's integer type, memcpy would do the exactly the same thing, right?

File: src/backend/access/transam/distributedlog.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (paul-guo-) [reply src/backend/access/transam/distributedlog.c]: I saw `DistributedLog_CommittedCheck()` has similar code that has two locks protected. Also please remove the redundant code using this function.

```
     LWLockAcquire(DistributedLogTruncateLock, LW_SHARED);
      LWLockAcquire(DistributedLogControlLock, LW_EXCLUSIVE);
```

**Comment by xiong-gang** [comment]
@pivotal-ning-yu could you please help review this.

File: src/test/isolation2/sql/concurrent_update.sql (hunk 7 above)
Comment (pivotal-ning-yu) [reply src/test/isolation2/sql/concurrent_update.sql]: > > We'd better put a call to `gp_wait_until_triggered_fault()` to ensure the suspending, otherwise there might be a chance that con2 `END` before con1 on the segment.
> 
> It triggers the injector when executing `END` and then block, I don't think we got the chance to call `gp_wait_until_triggered_fault`

This is the perfect scenario to use `gp_wait_until_triggered_fault()`.  In con1 we call `select gp_inject_fault('before_xact_end_procarray'...)`, in other con we call `gp_wait_until_triggered_fault()` to ensure that con1 has reached the fault.

> > BTW, we could use the short form of the fault injector, `gp_inject_fault(faultname text, type text, db_id int4)`, the other fields are filled with the same values with your long form.
> > And in new code it's recommended to call `gp_inject_fault()` like this:
> > ```sql
> > 1: select gp_inject_fault('before_xact_end_procarray', 'suspend', dbid) from gp_segment_configuration where content=0 and role='p';
> > ```
> 
> Since the other session will also hit the injector `before_xact_end_procarray` and we don't want it get suspended, we need to specify the number of occurrences, so can't use the simplified version

The short version is defined as below:

```sql
CREATE FUNCTION gp_inject_fault(
  faultname text,
  type text,
  db_id int4)
RETURNS text
AS $$ select gp_inject_fault($1, $2, '', '', '', 1, 1, 0, $3) $$
LANGUAGE SQL;
```

It passes the same arguments with your code, and it's easier for the readers to understand the code.

On the other hand, even if we use the long form (I'm totally fine with it), we'd better get the dbid with a join with `gp_segment_configuration` instead of hard coded values.  It helps us to run correctly when the mirror is activated.

File: src/backend/cdb/cdbtm.c (hunk 8 above)
Comment (pivotal-ning-yu) [reply src/backend/cdb/cdbtm.c]: Using `memcpy()` could be faster, even if it's not (thanks to the smart modern compilers), it generates fewer instructions than a plain or unlooped for-loop, so I'd +1 to @kainwen .

File: src/backend/access/transam/distributedlog.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/access/transam/distributedlog.c]: > I saw `DistributedLog_CommittedCheck()` has similar code that has two locks protected. Also please remove the redundant code using this function.
> 
> ```
>      LWLockAcquire(DistributedLogTruncateLock, LW_SHARED);
>       LWLockAcquire(DistributedLogControlLock, LW_EXCLUSIVE);
> ```

I also saw the lock being used in other places, that was why I asked.

File: src/backend/cdb/cdbtm.c [hunk 10]
```diff
@@ -307,6 +311,7 @@ notifyCommittedDtxTransaction(void)
 	Assert(Gp_role == GP_ROLE_DISPATCH);
 	Assert(DistributedTransactionContext == DTX_CONTEXT_QD_DISTRIBUTED_CAPABLE);
 	Assert(isCurrentDtxActivated());
+	ListCell   *l;
```
Comment (pivotal-ning-yu) [inline src/backend/cdb/cdbtm.c]: Let's move the variable declaration before the assertions.

File: src/backend/access/transam/distributedlog.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/access/transam/distributedlog.c]: Do you mean DistributedLogTruncateLock?

[the remark this answers is not in the document]
Comment (pivotal-ning-yu) [reply src/backend/access/transam/distributedlog.c]: In the current master, all access to `DistributedLogCtl->shared->page_buffer` are performed with the `DistributedLogControlLock` held, so I wonder we should hold the lock too.

However I'm not sure what are the purposes for `DistributedLogTruncateLock` and `DistributedLogControlLock`, when should use which.  Do you have some details?

File: src/test/isolation2/sql/concurrent_update.sql [hunk 11]
```diff
@@ -10,5 +10,51 @@ INSERT INTO t_concurrent_update VALUES(1,1,'test');
...
+2: UPDATE t_concurrent_update SET b=b+10 WHERE a=1;
+3: BEGIN;
+-- transaction 3 will wait transaction 1 on the segment
+3&: UPDATE t_concurrent_update SET b=b+10 WHERE a=1;
+
+-- transaction 2 suspend before commit, but it will wake up transaction 3 on segment
+2: select gp_inject_fault('before_xact_end_procarray', 'suspend', dbid) FROM gp_segment_configuration WHERE role='p' AND content=-1;
+1&: select gp_wait_until_triggered_fault('before_xact_end_procarray', 1, dbid) FROM gp_segment_configuration WHERE role='p' AND content=-1;
+2&: END;
+1<:
```
Comment (pivotal-ning-yu) [inline src/test/isolation2/sql/concurrent_update.sql]: I might be wrong, but will it be better to write this part like this?

```sql
2: select gp_inject_fault('before_xact_end_procarray', 'suspend', dbid) FROM gp_segment_configuration WHERE role='p' AND content=-1;
2&: END;
1: select gp_wait_until_triggered_fault('before_xact_end_procarray', 1, dbid) FROM gp_segment_configuration WHERE role='p' AND content=-1;
```

File: src/backend/access/transam/distributedlog.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (xiong-gang) [reply src/backend/access/transam/distributedlog.c]: `SimpleLruReadPage_ReadOnly` will get that lock
